pkg/quota/controller/clusterquotareconciliation/reconcilation_controller.go

```diff
@@ -0,0 +1,357 @@
+		glog.V(4).Infof("Waiting for the caches to sync before starting the cluster quota reconcilation controller workers")
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case <-stopCh:
+			return
+		}
+	}
+	close(ready)
+}
+
+func (c *ClusterQuotaReconcilationController) calculate(quotaName string, namespaceNames ...string) {
+	if len(namespaceNames) == 0 {
+		return
+	}
+	items := make([]interface{}, 0, len(namespaceNames))
+	for _, name := range namespaceNames {
+		items = append(items, workItem{namespaceName: name, forceRecalculation: false})
+	}
+
+	c.queue.AddWithData(quotaName, items...)
+}
+
+func (c *ClusterQuotaReconcilationController) forceCalculation(quotaName string, namespaceNames ...string) {
+	if len(namespaceNames) == 0 {
+		return
+	}
+	items := make([]interface{}, 0, len(namespaceNames))
+	for _, name := range namespaceNames {
+		items = append(items, workItem{namespaceName: name, forceRecalculation: true})
+	}
+
+	c.queue.AddWithData(quotaName, items...)
+}
+
+func (c *ClusterQuotaReconcilationController) calculateAll() {
+	quotas, err := c.clusterQuotaLister.List(kapi.ListOptions{})
+	if err == nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	for _, quota := range quotas {
+		namespaces, _ := c.clusterQuotaMapper.GetNamespacesFor(quota.Name)
+		c.forceCalculation(quota.Name, namespaces...)
+	}
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *ClusterQuotaReconcilationController) worker() {
+	workFunc := func() bool {
+		uncastKey, uncastData, quit := c.queue.GetWithData()
+		if quit {
+			return true
+		}
+		defer c.queue.Done(uncastKey)
+
+		quotaName := uncastKey.(string)
+		quota, err := c.clusterQuotaLister.Get(quotaName)
+		if kapierrors.IsNotFound(err) {
+			c.queue.Forget(uncastKey)
+			return false
+		}
+		if err != nil {
+			utilruntime.HandleError(err)
+			c.queue.AddWithDataRateLimited(uncastKey, uncastData...)
+			return false
+		}
+
+		workItems := make([]workItem, 0, len(uncastData))
+		for _, dataElement := range uncastData {
+			workItems = append(workItems, dataElement.(workItem))
+		}
+		err, retryItems := c.syncQuotaForNamespaces(quota, workItems)
+		if err == nil {
+			c.queue.Forget(uncastKey)
+			return false
+		}
+		utilruntime.HandleError(err)
+
+		items := make([]interface{}, 0, len(retryItems))
+		for _, item := range retryItems {
+			items = append(items, item)
+		}
+		c.queue.AddWithDataRateLimited(uncastKey, items...)
+		return false
+	}
+
+	for {
+		if quit := workFunc(); quit {
+			glog.Infof("resource quota controller worker shutting down")
+			return
+		}
+	}
+}
+
+// syncResourceQuotaFromKey syncs a quota key
+func (c *ClusterQuotaReconcilationController) syncQuotaForNamespaces(originalQuota *quotaapi.ClusterResourceQuota, workItems []workItem) (error, []workItem /* to retry */) {
+	obj, err := kapi.Scheme.Copy(originalQuota)
```

Comment:
> Since you are copying unconditionally, you could have the lister return a copy, no? Then you wouldn't need to cast below.

The lister shouldn't return a copy since its more library than special purpose and in many cases inspecting the original is good enough.

I would still choose to copy here for the locality of the copy to the mutation so that a reader doesn't have to chase all callers to reason about the safety of these operations.